tests/unit/test_cornac_utils.py

```diff
@@ -0,0 +1,136 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+
+import pandas as pd
+import pytest
+import cornac
+
+from reco_utils.common.constants import (
+    DEFAULT_USER_COL,
+    DEFAULT_ITEM_COL,
+    DEFAULT_RATING_COL,
+)
+from reco_utils.recommender.cornac.cornac_utils import (
+    predict_rating,
+    predict_ranking,
+)
+
+TOL = 0.001
+
+
+@pytest.fixture
+def rating_true():
+    return pd.DataFrame(
+        {
+            DEFAULT_USER_COL: [1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3],
+            DEFAULT_ITEM_COL: [
+                1,
+                2,
+                3,
+                1,
+                4,
+                5,
+                6,
+                7,
+                2,
+                5,
+                6,
+                8,
+                9,
+                10,
+                11,
+                12,
+                13,
+                14,
+            ],
+            DEFAULT_RATING_COL: [5, 4, 3, 5, 5, 3, 3, 1, 5, 5, 5, 4, 4, 3, 3, 3, 2, 1],
+        }
+    )
+
+
+def test_predict_rating(rating_true):
+    mf = cornac.models.MF()
+    train_set = cornac.data.Dataset.from_uir(rating_true.itertuples(index=False), seed=42)
+    mf.fit(train_set)
+
+    preds = predict_rating(mf, rating_true)
+    assert set(preds.columns) == {"userID", "itemID", "prediction"}
+    assert preds["userID"].dtypes == rating_true["userID"].dtypes
+    assert preds["itemID"].dtypes == rating_true["itemID"].dtypes
+    user = rating_true.iloc[0]["userID"]
+    item = rating_true.iloc[0]["itemID"]
+    assert preds[(preds["userID"] == user) & (preds["itemID"] == item)][
+               "prediction"
+           ].values == pytest.approx(mf.rate(train_set.uid_map[user],
+                                             train_set.iid_map[item]).item(),
+                                     rel=TOL)
+
+    preds = predict_rating(
+        mf,
+        rating_true,
+        usercol="userID",
+        itemcol="itemID",
+        predcol="prediction",
+    )
+    assert set(preds.columns) == {"userID", "itemID", "prediction"}
+    assert preds["userID"].dtypes == rating_true["userID"].dtypes
+    assert preds["itemID"].dtypes == rating_true["itemID"].dtypes
+    user = rating_true.iloc[1]["userID"]
+    item = rating_true.iloc[1]["itemID"]
+    assert preds[(preds["userID"] == user) & (preds["itemID"] == item)][
+               "prediction"
+           ].values == pytest.approx(mf.rate(train_set.uid_map[user],
```

Comment:
here we are cheating :-) because we are comparing two routines that are the same. What we want is to compare the prediction that comes from `predict_rating` with a df explicitly created by us as a fixture, so you would need to generate a new df with the result.

Question, any reason to select MF instead of BPR for the test?

Reply:
I think this test is still valid. We are testing ground-truth training data and the predictions from a trained MF model. In theory, MF should be able to overfit on this small input data, thus, it will be a small approximation error here.

Reply:
MF is used instead of BPR because I want to reuse tests from Surprise (MF is equivalent to SVD in Surprise). For the purpose of testing utility functions, it should be fine.

Reply:
Hi @miguelgfierro ,

Tests are updated using some evaluation metrics. It might not be that clean since it involves the implementation of metrics. However, I think it's reliable enough to be unit tests.